File: go/recursed/collatz.go
package main

import "fmt"
import "sort"

func sequenceLength(num int) int {
    length := 0
    for num > 1    {
        if num%2 == 0        {
            num = num / 2
        }        else        {
            num = num * 3 + 1
        }
        length += 1
    }
    return length
}

//a custom elem type that contains a num and it's sequence length
type elem struct {
    num int
    length  int
}

//compare elems first by length then by num
func compareElems(e1, e2 elem) bool {
    if e1.length == e2.length {
        return e1.num > e2.num
    }    else {
        return e1.length < e2.length
    }
}

//all these things are need to sort with the custom type elem
type byLength []elem

func (s byLength) Len() int {
    return len(s)
}
func (s byLength) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}
func (s byLength) Less(i, j int) bool {
    return compareElems(s[i], s[j])
}

func main() {

   resultArray := make([]elem, 10)
   myNum := 9
   myElem := elem{myNum,sequenceLength(myNum)}
   fmt.Printf("%d has a length %d\n", myElem.num, myElem.length)

   resultArray[5] = elem{-1,-1}

   sort.Sort(byLength(resultArray))

   fmt.Println(resultArray)

}

File: julia/recursed/collatz.jl
#!/usr/bin/julia
import Base.isless

#recursively calculate the sequence length of a number
function sequenceLength(num::Integer)
    if num == 1
        return 0
    elseif num%2 == 0
        return 1 + sequenceLength(num ÷ 2)
    else
        return 1 + sequenceLength(num * 3 + 1)
    end
end

#a custom data type that has a number and it's sequence length
struct elem
    num
    length
end

#override isless for our custom data type
function isless(a::elem, b::elem)
    if a.length == b.length
        return a.num > b.num
    else
        return a.length < b.length
    end
end

function main()
    result = fill(elem(0,0), 10)

    for i = 1:4999999999999
        newElem = elem(i, sequenceLength(i))
        minIndex = argmin(result)
        if isless(result[minIndex], newElem)
            result[minIndex] = newElem
        end
    end

    sort!(result)

    for e in result
        println("$(e.num) has a length of $(e.length)")
    end
end

main()
